### PR TITLE
Feature/periodically scan disk

### PR DESCRIPTION
**Pull request by parkerduckworth**
### Changes

Introduce 2 new configuration variables:

##### `DISK_USE_WARNING_PERCENTAGE`:
  - Sets disk use threshold for triggering log warnings

##### `DISK_USE_READONLY_PERCENTAGE`:
  - Sets disk use threshold for triggering shard status set to `READONLY`

### CODE REVIEW

File: usecases/config/config_handler.go [hunk 1]
```diff
@@ -118,6 +119,25 @@ func (p Persistence) Validate() error {
 	return nil
 }
 
+// TODO: Note the mix of camelCase and snake_case
```
Comment (antas-marcin) [inline usecases/config/config_handler.go:122]: remove TODO comment

File: usecases/config/config_handler.go [hunk 2]
```diff
@@ -118,6 +119,25 @@ func (p Persistence) Validate() error {
...
+type DiskUse struct {
+	WarningPercentage  uint64 `json:"warning_percentage" yaml:"warning_percentage"`
+	ReadOnlyPercentage uint64 `json:"readonly_percentage" yaml:"readonly_percentage"`
+}
+
+// TODO: Does it make sense to allow for setting up to 100%?
```
Comment (antas-marcin) [inline usecases/config/config_handler.go:128]: here also I think this TODO comment can be remvoed

File: usecases/config/config_handler.go (hunk 1 above)
Comment (parkerduckworth) [reply usecases/config/config_handler.go:122]: Ah yes this reminds me that I need to raise this issue. Thanks

File: adapters/repos/db/shard.go [hunk 3]
```diff
@@ -173,6 +173,8 @@ func (s *Shard) drop(force bool) error {
 		return storagestate.ErrStatusReadOnly
 	}
 
+	s.cancel <- struct{}{}
```
[the remark this answers is not in the document]
Comment (parkerduckworth) [reply adapters/repos/db/shard.go:176]: is there a reason to continue scanning if the shard has been dropped?

File: adapters/repos/db/shard.go [hunk 4]
```diff
@@ -72,8 +72,8 @@ func NewShard(ctx context.Context, shardName string, index *Index) (*Shard, erro
 		deletedDocIDs:    docid.NewInMemDeletedTracker(),
 		cleanupInterval: time.Duration(index.invertedIndexConfig.
 			CleanupIntervalSeconds) * time.Second,
-		cleanupCancel: make(chan struct{}),
-		randomSource:  rand,
+		cancel:       make(chan struct{}, 1),
```
Comment (antas-marcin) [inline adapters/repos/db/shard.go:75]: why are you creating a buffered channel?

Comment (parkerduckworth) [reply adapters/repos/db/shard.go:75]: If it is not buffered, the shutdown/drop is blocked indefinitely if there is no receiver. Right now the disk use scanner is the only receiver, so if the scan is not initiated (for example during a test, or if the user does not set the disk use config options), shutdown/drop cannot execute to completion.

Comment (parkerduckworth) [reply adapters/repos/db/shard.go:75]: i actually had forgotten to add the logic to only scan if use explicitly passes the config to do so. just pushed up a moment ago. please let me know if you have any specific issues with creating a buffered channel here, or want to discuss more about it

**Pull request merged**